nft-token-factory/src/lib.rs

```diff
@@ -0,0 +1,627 @@
+    pub expires_at: Option<String>, // ISO 8601 datetime when token expires
+    pub starts_at: Option<String>, // ISO 8601 datetime when token starts being valid
+    pub updated_at: Option<String>, // ISO 8601 datetime when token was last updated
+    pub extra: Option<String>, // anything extra the NFT wants to store on-chain. Can be stringified JSON.
+    pub reference: Option<String>, // URL to an off-chain JSON file with more info.
+    pub reference_hash: Option<Base64VecU8>, // Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
+}
+
+#[ext_contract(ext_bridge_nft)]
+pub trait ExtBridgedNFT {
+    fn nft_mint(&self, token_id: String, recipient: AccountId, metadata: TokenMetadata);
+    fn set_metadata(&mut self, metadata: NFTMetadata);
+    fn set_owner_account_id(&mut self, new_owner: ValidAccountId);
+}
+
+#[near_bindgen]
+impl NFTFactory {
+    #[init]
+    pub fn new(prover_account: AccountId, locker_address: String) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        Self {
+            prover_account,
+            locker_address: get_eth_address(locker_address),
+            tokens: UnorderedSet::new(b"t".to_vec()),
+            used_events: UnorderedSet::new(b"u".to_vec()),
+            owner_pk: env::signer_account_pk(),
+            owner_id: env::signer_account_id(),
+            bridge_token_storage_deposit_required: env::storage_byte_cost() * 237, // this can be verified by calling storage_cost_per_nft() in bridge_nft.wasm when deployed
+            paused: Mask::default(),
+        }
+    }
+
+    /// Return all registered tokens
+    pub fn get_tokens(&self) -> Vec<String> {
+        self.tokens.iter().collect::<Vec<_>>()
+    }
+
+    /// Finalise the withdraw to eth from a sub NFT contract. Only this bridge can emit the execution outcome to be processed on the Eth side
+    /// Caller must be <token_address>.<current_account_id>, where <token_address> exists in the `tokens`.
+    // todo: how much GAS is required to execute this method
+    #[result_serializer(borsh)]
+    pub fn finish_withdraw_to_eth(
+        &mut self,
+        #[serializer(borsh)] token_id: String,
+        #[serializer(borsh)] recipient: String,
+    ) -> ResultType {
+        self.check_not_paused(PAUSE_NEAR_TO_ETH_TRANSFER);
+
+        let token = env::predecessor_account_id();
+
+        let parts: Vec<&str> = token.split(".").collect();
+        assert_eq!(
+            token,
+            format!("{}.{}", parts[0], env::current_account_id()),
+            "Only sub accounts of NFT Factory can call this method."
+        );
+
+        assert!(
+            self.tokens.contains(&parts[0].to_string()),
+            "Such Bridge NFT token does not exist."
+        );
+
+        let token_address = get_eth_address(parts[0].to_string());
+        let recipient_address = get_eth_address(recipient);
```

Comment:
This two functions needs to run on the bridged-nft-token side, given this function must not fail.

Reply:
Fixed [here ](https://github.com/aurora-is-near/rainbow-non-fungible-token-connector/pull/4/commits/e64010864c745d7520b1dc182529a501f95c89f4)